### PR TITLE
Allow message_progress_bar to be a function

### PR DESCRIPTION
**Pull request by Shaobin-Jiang**
I have encountered this situation many times when I want to update the progress bar message when the progress bar updates, but the only way of doing it before is manually changing the `innerHTML` in an `on_finish` function of a trial. Therefore, I think we should allow users to set `message_progress_bar` as a function that returns the progress, so that the message gets updated automatically.

For example:

```javascript
const jsPsych = initJsPsych({
    show_progress_bar: true,
    auto_update_progress_bar: true,
    message_progress_bar: (progress) => `Current progress: ${progress * 100}%`
});
```

### CODE REVIEW

**Comment by jodeleeuw** [comment]
Thanks @Shaobin-Jiang! This seems useful to me.

@jspsych/core what do you think about having `message_progress_bar` be a string or function? Should it just be a function?

**Comment by Shaobin-Jiang** [comment]
I personally think `message_progress_bar` could either be a string or a function. There must be many who are happy with displaying just an unchanging message along side with the progress bar. Also I believe in this way `message_progress_bar` can behave like a dynamic parameter.

**Comment by bjoluc** [comment]
I tend to agree with @Shaobin-Jiang. Also, supporting both really doesn't add much complexity implementation-wise.

File: .changeset/stupid-baboons-wait.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"jspsych": major
```
Comment (bjoluc) [inline .changeset/stupid-baboons-wait.md:2]:
```suggestion
"jspsych": minor
```
Not a breaking change, just a feature.

File: packages/jspsych/src/ProgressBar.ts [hunk 2]
```diff
@@ -14,7 +17,13 @@ export class ProgressBar {
   /** Adds the progress bar HTML code into `this.containerElement` */
   private setupElements() {
     this.messageSpan = document.createElement("span");
-    this.messageSpan.innerHTML = this.message;
+
+    if (Object.prototype.toString.call(this.message) === "[object Function]") {
```
Comment (bjoluc) [inline packages/jspsych/src/ProgressBar.ts:21]: Why not use `typeof this.message === "function"`?

File: packages/jspsych/src/ProgressBar.ts [hunk 3]
```diff
@@ -14,7 +17,13 @@ export class ProgressBar {
...
+      this.messageSpan.innerHTML = (this.message as (progress: number) => string)(0);
+    } else {
+      // Progress starts at 0 when experiment commences
+      this.messageSpan.innerHTML = this.message as string;
+    }
```
Comment (bjoluc) [inline packages/jspsych/src/ProgressBar.ts:26]: I'd prefer extracting this bit into a private `updateMessage()` method so we don't need to duplicate parts of it in the `update()` method.

File: .changeset/stupid-baboons-wait.md (hunk 1 above)
Comment (Shaobin-Jiang) [reply .changeset/stupid-baboons-wait.md]: Done in commit 9727773.

File: packages/jspsych/src/ProgressBar.ts (hunk 2 above)
Comment (Shaobin-Jiang) [reply packages/jspsych/src/ProgressBar.ts]: My mistake. I think I mixed it up with something else.

File: packages/jspsych/src/ProgressBar.ts (hunk 3 above)
Comment (Shaobin-Jiang) [reply packages/jspsych/src/ProgressBar.ts]: Fixed this and also the function type check in 4dbdbc0.

File: packages/jspsych/src/ProgressBar.ts [hunk 4]
```diff
@@ -31,6 +35,16 @@ export class ProgressBar {
   /** Updates the progress bar according to `this.progress` */
   private update() {
     this.innerDiv.style.width = this._progress * 100 + "%";
+
+    this.updateMessage();
+  }
+
+  private updateMessage() {
```
Comment (bjoluc) [inline packages/jspsych/src/ProgressBar.ts:42]: @Shaobin-Jiang I missed that we always need this together with `update()` anyway – would you mind just joining it into the existing `update` function? We're already calling update in `setupElements()` too... Thanks for all the changes and the unit test – looks very good!

Comment (Shaobin-Jiang) [reply packages/jspsych/src/ProgressBar.ts]: Done.

**Review by bjoluc — APPROVED** [review]

**Pull request merged**